# 0.0 Imports

In [1]:
import json
import rarfile
import warnings

import numpy   as np
import pandas  as pd
import seaborn as sns

from pymongo import MongoClient

from sklearn import cluster       as cl
from sklearn import metrics       as mt
from sklearn import preprocessing as pp

# 0.2 Loading Data

In [2]:
df_raw = pd.read_csv('C:/Users/PICHAU/repos/cluster_project/data/raw/data.csv', encoding = 'unicode_escape') 

# 1.0 Data Description

In [3]:
df1 = df_raw.copy()

In [4]:
df1.sample(5)

InvoiceNo StockCode                         Description  Quantity  \
276315    561038     23066          SMALL DECO JEWELLERY STAND         1   
321713    565198     22558     CLOTHES PEGS RETROSPOT PACK 24          2   
225146    556636     22713                 CARD I LOVE LONDON          2   
73875     542393     22969        HOMEMADE JAM SCENTED CANDLES        24   
162810    550519    85123A  WHITE HANGING HEART T-LIGHT HOLDER        18   

            InvoiceDate  UnitPrice  CustomerID         Country  
276315  7/24/2011 11:58       8.25     17114.0  United Kingdom  
321713   9/1/2011 16:18       1.65     16729.0  United Kingdom  
225146  6/13/2011 15:30       0.83         NaN  United Kingdom  
73875   1/27/2011 15:02       1.45     18125.0  United Kingdom  
162810   4/19/2011 9:31       2.95     13141.0  United Kingdom

## 1.1 Data Dimensions and Types

In [5]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


## 1.2 Check NaN's

In [6]:
df1.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

#### 1.2.1.2 Excluding NaN's

In [7]:
df1 = df1.dropna(subset=['Description', 'CustomerID'])

In [8]:
df1.isna().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [9]:
print('Removed data: {:.0%}'.format( 1-(df1.shape[0] / df_raw.shape[0])))
print(f'New number of rows: {df1.shape[0]}')

Removed data: 25%
New number of rows: 406829


## 1.3 Change Types

In [10]:
df1['InvoiceDate'] = pd.to_datetime(df1['InvoiceDate'])
df1['CustomerID'] = df1['CustomerID'].astype('int64')

In [11]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 406829 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    406829 non-null  object        
 1   StockCode    406829 non-null  object        
 2   Description  406829 non-null  object        
 3   Quantity     406829 non-null  int64         
 4   InvoiceDate  406829 non-null  datetime64[ns]
 5   UnitPrice    406829 non-null  float64       
 6   CustomerID   406829 non-null  int64         
 7   Country      406829 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 27.9+ MB


## 1.4 Descritive Statistics

## 1.4.1 Numerical Statistics

In [12]:
num_attributes = df1.select_dtypes(include=['int64', 'float64'])

# Central Tendency - mean, median
ct1 = pd.DataFrame(num_attributes.apply(np.mean)).T
ct2 = pd.DataFrame(num_attributes.apply(np.median)).T

# dispersion - std, min, max, range, skew, kurtosis
d1 = pd.DataFrame(num_attributes.apply(np.std)).T
d2 = pd.DataFrame(num_attributes.apply(min)).T
d3 = pd.DataFrame(num_attributes.apply(max)).T
d4 = pd.DataFrame(num_attributes.apply(lambda x: x.max() - x.min())).T
d5 = pd.DataFrame(num_attributes.apply(lambda x: x.skew())).T
d6 = pd.DataFrame(num_attributes.apply(lambda x: x.kurtosis())).T

# concatenar
m = pd.concat([d2, d3, d4, ct1, ct2, d1, d5, d6]).T.reset_index()
m.columns = ['attributes', 'min', 'max', 'range', 'mean', 'median', 'std', 'skew', 'kurtosis']
m

attributes      min      max     range          mean    median  \
0    Quantity -80995.0  80995.0  161990.0     12.061303      5.00   
1   UnitPrice      0.0  38970.0   38970.0      3.460471      1.95   
2  CustomerID  12346.0  18287.0    5941.0  15287.690570  15152.00   

           std        skew       kurtosis  
0   248.693064    0.182663   94317.563673  
1    69.315077  452.219019  246924.542988  
2  1713.598197    0.029835      -1.179982

<b>INVESTIGATING</b>:

== Quantity values of max and min;

In [13]:
ds = df1[(df1['Quantity']<0)]
print('Number of rows that Quantity is negative: ', len(ds))

Number of rows that Quantity is negative:  8905


## 1.4.2 Categorical Statistics

In [14]:
cat_att = df1.select_dtypes(exclude=['int64', 'float64', 'datetime64[ns]'])

In [15]:
# invoice number
cs = cat_att[(cat_att['InvoiceNo'].str.contains('[^0-9]+', regex=True))]
print('Number of rows that invoice number has C: ', len(cs))

Number of rows that invoice number has C:  8905


In [16]:
# stock
cs2 = cat_att[(cat_att['StockCode'].str.contains('^[a-zA-Z\s]+$', regex=True))]
print('Stock code values out: {}'.format(cs2['StockCode'].unique()))

Stock code values out: ['POST' 'D' 'M' 'BANK CHARGES' 'PADS' 'DOT' 'CRUK']


In [17]:
# Country
cs3 = df1[['CustomerID', 'Country']].drop_duplicates().groupby('Country').count().reset_index()
cs3

Country  CustomerID
0              Australia           9
1                Austria          11
2                Bahrain           2
3                Belgium          25
4                 Brazil           1
5                 Canada           4
6        Channel Islands           9
7                 Cyprus           8
8         Czech Republic           1
9                Denmark           9
10                  EIRE           3
11    European Community           1
12               Finland          12
13                France          87
14               Germany          95
15                Greece           4
16               Iceland           1
17                Israel           4
18                 Italy          15
19                 Japan           8
20               Lebanon           1
21             Lithuania           1
22                 Malta           2
23           Netherlands           9
24                Norway          10
25                Poland           6
26              Portugal          19
27                   RSA           1
28          Saudi Arabia           1
29             Singapore           1
30                 Spain          31
31                Sweden           8
32           Switzerland          21
33                   USA           4
34  United Arab Emirates           2
35        United Kingdom        3950
36           Unspecified           4

# 2.0 Data Filtration

In [18]:
df2 = df1.copy()

In [19]:
 # Numerical attribute 
df2 = df2[(df2['UnitPrice'] >= 0.05)]

# Categorical attributes
df2 = df2[~df2['StockCode'].isin(['POST' 'D' 'M' 'BANK CHARGES' 'PADS' 'DOT' 'CRUK'])]

# --- Description
df2 = df2.drop(columns='Description', axis=1)

# --- Country 
df2 = df2[~df2['Country'].isin(['European Community', 'Unspecified'])]

In [20]:
print('Number of rows after filtration: ',len(df2))

Number of rows after filtration:  406410


# 3.0 Feature Engineering

## 3.1 Feature Creation

In [21]:
df2['GrossRevenue'] = df2['Quantity'] * df2['UnitPrice']

In [22]:
df3 = df2.copy()

In [23]:
# refference table - customer
df_ref = pd.DataFrame(df3['CustomerID']).drop_duplicates(keep='first', ignore_index=True)
#df_ref = df_ref.dropna(subset=['CustomerID'])

In [24]:
# Gross Revenue (Monetary)
df_monetary = df3[['GrossRevenue','CustomerID']].groupby('CustomerID').sum().reset_index()
df_ref = pd.merge(df_ref, df_monetary, on='CustomerID', how='left')

In [25]:
# Recency days (Recency)
df_recency = df3[['CustomerID', 'InvoiceDate']].groupby('CustomerID').max().reset_index()
df_recency['RecencyDays'] = (df3['InvoiceDate'].max() - df_recency['InvoiceDate']).dt.days
df_recency = df_recency[['CustomerID', 'RecencyDays']]
df_ref = pd.merge(df_ref, df_recency, on='CustomerID', how='left')

In [26]:
# Purchase number per day (Frequency)
df_aux = (df3[['CustomerID', 'InvoiceNo', 'InvoiceDate']].drop_duplicates()
                                                             .groupby('CustomerID')
                                                             .agg( max_ = ('InvoiceDate', 'max'), 
                                                                   min_ = ('InvoiceDate', 'min'),
                                                                   days_= ('InvoiceDate', lambda x: ((x.max() - x.min()).days ) + 1 ),
                                                                   buy_ = ('InvoiceNo', 'count'))).reset_index()
# Frequency
df_aux['Frequency'] = df_aux[['buy_', 'days_']].apply( lambda x: x['buy_'] / x['days_'] if  x['days_'] != 0 else 0, axis=1 )

# Merge
df_ref = pd.merge(df_ref, df_aux[['CustomerID', 'Frequency']], on='CustomerID', how='left')

In [27]:
# ReturnsNo
df_in2 = df3[df3['InvoiceNo'].str.contains('[^0-9]+', regex=True)]
df_rn = df_in2[['CustomerID', 'InvoiceNo']].drop_duplicates().groupby('CustomerID').count().reset_index().rename(columns={'InvoiceNo':'ReturnsNo'})
df_ref = pd.merge(df_ref, df_rn, on='CustomerID', how='left')

In [28]:
# Products Quantity
df_freq = (df3.loc[:, ['CustomerID', 'Quantity']].groupby('CustomerID').sum()
                                                           .reset_index()
                                                           .rename(columns={'Quantity': 'QtItems'}))
df_ref = pd.merge( df_ref, df_freq, on='CustomerID', how='left')

In [29]:
# Product Returns
df_in3 = df3[df3['InvoiceNo'].str.contains('[^0-9]+', regex=True)]
df_pr= df_in3[['CustomerID', 'Quantity']].groupby('CustomerID').sum().reset_index().rename(columns={'Quantity':'ProductReturns'})
df_ref = pd.merge(df_ref, df_pr, on='CustomerID', how='left')

# --- Converting negative values for positive 
df_ref['ProductReturns'] = df_ref['ProductReturns'] * -1

In [30]:
# Purchase number
df_in4 = df3[~df3['InvoiceNo'].str.contains('[^0-9]+', regex=True)]
df_pn = df_in4[['CustomerID', 'InvoiceNo']].drop_duplicates().groupby('CustomerID').count().reset_index().rename(columns={'InvoiceNo':'PurchasesNo'})
df_ref = pd.merge(df_ref, df_pn, on='CustomerID', how='left')

In [31]:
# Products number
df_pn1 = df3[~df3['InvoiceNo'].str.contains('[^0-9]+', regex=True)]
df_pn2 = (df_pn1[['CustomerID', 'StockCode']].groupby('CustomerID')
                                                      .count()
                                                      .reset_index()
                                                      .rename(columns={'StockCode':'ProductsNo'}))

df_ref = pd.merge(df_ref, df_pn2, on = 'CustomerID', how = 'left')

In [32]:
# Purchasing month
df3['InvoiceMonth'] = df3['InvoiceDate'].dt.month
df_r = df3[['CustomerID', 'InvoiceMonth']].drop_duplicates().groupby('CustomerID').count().reset_index().rename(columns={'InvoiceMonth':'PurchasingMonths'})

df_ref = pd.merge(df_ref, df_r[['CustomerID', 'PurchasingMonths']], on = 'CustomerID', how = 'left')

In [33]:
# the result table
df_ref

CustomerID  GrossRevenue  RecencyDays  Frequency  ReturnsNo  QtItems  \
0          17850       5288.63          301   0.486111        1.0     1693   
1          13047       3079.10           31   0.052478        8.0     1355   
2          12583       7187.34            2   0.048518        3.0     5009   
3          13748        948.25           95   0.017921        NaN      439   
4          15100        635.10          329   0.136364        3.0       58   
...          ...           ...          ...        ...        ...      ...   
4361       13436        196.89            1   1.000000        NaN       76   
4362       15520        343.50            1   1.000000        NaN      314   
4363       13298        360.00            0   1.000000        NaN       96   
4364       14569        227.39            0   1.000000        NaN       79   
4365       12713        848.55            0   1.000000        NaN      508   

      ProductReturns  PurchasesNo  ProductsNo  PurchasingMonths  
0               40.0         34.0       297.0                 2  
1               36.0         10.0       172.0                10  
2               51.0         15.0       247.0                11  
3                NaN          5.0        28.0                 3  
4               22.0          3.0         3.0                 2  
...              ...          ...         ...               ...  
4361             NaN          1.0        12.0                 1  
4362             NaN          1.0        18.0                 1  
4363             NaN          1.0         2.0                 1  
4364             NaN          1.0        12.0                 1  
4365             NaN          1.0        38.0                 1  

[4366 rows x 10 columns]

## 3.2 Check NaN's From Features Created

In [34]:
df_ref.isna().sum()

CustomerID             0
GrossRevenue           0
RecencyDays            0
Frequency              0
ReturnsNo           2778
QtItems                0
ProductReturns      2778
PurchasesNo           33
ProductsNo            33
PurchasingMonths       0
dtype: int64

## 3.3 Fill NaN's

In [35]:
df_ref['ReturnsNo'] = df_ref['ReturnsNo'].fillna(value=0)
df_ref['ProductReturns'] = df_ref['ProductReturns'].fillna(value=0)
df_ref['PurchasesNo'] = df_ref['PurchasesNo'].fillna(value=0)
df_ref['ProductsNo'] = df_ref['ProductsNo'].fillna(value=0)

In [36]:
df_ref.isna().sum()

CustomerID          0
GrossRevenue        0
RecencyDays         0
Frequency           0
ReturnsNo           0
QtItems             0
ProductReturns      0
PurchasesNo         0
ProductsNo          0
PurchasingMonths    0
dtype: int64

## 3.2 Data Filtration

In [37]:
# The gross revenue of some customers results in 0, this happened because they showed an atypical behavior (buying and returning the same quantities of products).
# They were considered bad users and filtered in this section.

df_ref = df_ref[df_ref['GrossRevenue']>=10]

In [38]:
df_ref.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4310 entries, 0 to 4365
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   CustomerID        4310 non-null   int64  
 1   GrossRevenue      4310 non-null   float64
 2   RecencyDays       4310 non-null   int64  
 3   Frequency         4310 non-null   float64
 4   ReturnsNo         4310 non-null   float64
 5   QtItems           4310 non-null   int64  
 6   ProductReturns    4310 non-null   float64
 7   PurchasesNo       4310 non-null   float64
 8   ProductsNo        4310 non-null   float64
 9   PurchasingMonths  4310 non-null   int64  
dtypes: float64(6), int64(4)
memory usage: 370.4 KB


# 4.0 EDA

In [39]:
df4 = df_ref.copy()

# 5.0  Data Preparation

In [40]:
df5 = df4.drop(columns=['CustomerID'], axis=1).copy()

In [41]:
mm = pp.MinMaxScaler()

df5['GrossRevenue'] = mm.fit_transform(df5[['GrossRevenue']])

df5['Frequency'] = mm.fit_transform(df5[['Frequency']])

df5['QtItems'] = mm.fit_transform(df5[['QtItems']])

df5['PurchasesNo'] = mm.fit_transform(df5[['PurchasesNo']])

df5['PurchasingMonths'] = mm.fit_transform(df5[['PurchasingMonths']])

# 6.0 Feature Selection

In [42]:
cols_selected = ['GrossRevenue', 'Frequency', 'QtItems', 'PurchasesNo', 'PurchasingMonths']

In [43]:
df6 = df5[cols_selected].copy()

# 7.0 Hyperparameter Fine Tunning

In [44]:
df7 = df6.copy()

# 8.0 Machine Learning

In [45]:
df8 = df6.copy()

## 8.1 K-Means

In [46]:
# Clusters
k=6

# Model
kmeansf = cl.KMeans(n_clusters=k, n_init=10, max_iter=300, random_state=32)
#init='random'/n_init=10 / random_state=42
kmeansf.fit(df8)

kflabels = kmeansf.labels_

### 8.1.2 Cluster Validation

In [47]:
# SS
print('Silhouette Score KMeans: ', mt.silhouette_score(df8, kflabels, metric='euclidean'))

Silhouette Score KMeans:  0.7133689169369299


# 9.0 Cluster Analysis

In [48]:
df9 = df8.copy()
df9['Clusters'] = kflabels

In [49]:
df9.head(5)

GrossRevenue  Frequency   QtItems  PurchasesNo  PurchasingMonths  Clusters
0      0.018878   0.096235  0.008631     0.158654          0.090909         3
1      0.010972   0.009413  0.006908     0.043269          0.818182         2
2      0.025672   0.008620  0.025537     0.067308          0.909091         2
3      0.003347   0.002494  0.002238     0.019231          0.181818         5
4      0.002227   0.026208  0.000296     0.009615          0.090909         3

## 9.1 Cluster Profile

In [50]:
df94 = df4.copy() 
df94['Clusters'] = kflabels
df94.sample(5)

CustomerID  GrossRevenue  RecencyDays  Frequency  ReturnsNo  QtItems  \
298        16411        214.65           18   0.005714        0.0      132   
1180       12422        803.56           95   0.013043        0.0      370   
4199       17397        696.41           21   2.000000        0.0      254   
3247       14116       1382.74           19   0.049020        0.0      707   
3905       15677        657.23           48   1.000000        0.0      319   

      ProductReturns  PurchasesNo  ProductsNo  PurchasingMonths  Clusters  
298              0.0          2.0        45.0                 2         3  
1180             0.0          3.0        39.0                 3         5  
4199             0.0          2.0        92.0                 1         0  
3247             0.0          5.0        72.0                 4         5  
3905             0.0          1.0        33.0                 1         0

In [51]:
# clusters composition by total database 
df_cluster = df94[['CustomerID', 'Clusters']].groupby('Clusters').count().reset_index()

In [52]:
df_cluster['Pencentage'] = 100 * (df_cluster['CustomerID'] / df_cluster['CustomerID'].sum())

# Gross revenue average
avg_rev = df94[['Clusters', 'GrossRevenue']].groupby('Clusters').mean().reset_index() 
df_cluster = pd.merge(df_cluster, avg_rev, on= 'Clusters', how='inner')

# Recency average
avg_rec = df94[['Clusters', 'RecencyDays']].groupby('Clusters').mean().reset_index() 
df_cluster = pd.merge(df_cluster, avg_rec, on= 'Clusters', how='inner')

# PurchasesNo average
avg_pn = df94[['Clusters', 'PurchasesNo']].groupby('Clusters').mean().reset_index() 
df_cluster = pd.merge(df_cluster, avg_pn, on= 'Clusters', how='inner')

# Frequency average
avg_t = df94[['Clusters', 'Frequency']].groupby('Clusters').mean().reset_index()
df_cluster = pd.merge(df_cluster, avg_t, on= 'Clusters', how='inner')

# PurchasinMonths average
avg_t = df94[['Clusters', 'PurchasingMonths']].groupby('Clusters').mean().reset_index() 
df_cluster = pd.merge(df_cluster, avg_t, on= 'Clusters', how='inner')

df_cluster.sort_values('GrossRevenue', ascending=False)

Clusters  CustomerID  Pencentage  GrossRevenue  RecencyDays  PurchasesNo  \
2         2         185    4.292343  17553.722486     7.772973    27.108108   
4         4         287    6.658933   4376.088084    18.648084    10.822300   
1         1         460   10.672854   2624.391391    33.402174     6.526087   
5         5         946   21.948956   1409.850909    55.287526     3.855180   
3         3        1034   23.990719    720.870803    93.559961     2.156673   
0         0        1398   32.436195    371.665901   152.829757     1.067954   

   Frequency  PurchasingMonths  
2   0.093617         10.972973  
4   0.042605          7.864111  
1   0.031402          5.384783  
5   0.030935          3.421776  
3   0.071932          1.865571  
0   1.060885          1.002861

<b>CLUSTER 2 (VEIGHERS): </b>

Total: 185 customers (4.3% of total customers)

Spent (avg): $ 17553.72 

Recency (avg):  8 days

Purchase (avg): 27 purchases

Purchases per day (avg): 0.09 products/day

Purchasing Months (avg): 11 months 

-------------------------------------------------

<b>CLUSTER 4 (ASCENDING): </b>

Total: 287 customers (6.7% of total customers)

Spent (avg): $ 4376.09  

Recency (avg): 19 days

Purchase (avg): 11 purchases

Purchases per day (avg): 0.04 products/day

Purchasing Months (avg): 8 months  

-------------------------------------------------

<b>CLUSTER 1 (IMPROVE): </b>

Total: 460 customers (10.67% of total customers)

Spent (avg): $ 2624.40 

Recency (avg): 34 days

Purchase (avg): 6 purchases

Purchases per day (avg): 0.03 products/day

Purchasing Months (avg): 6 months 

-------------------------------------------------

<b>CLUSTER 5 (RESCUE): </b>

Total: 946 customers (21.94% of total customers)

Spent (avg): $ 1409.85

Recency (avg): 56 days

Purchase (avg): 4 purchases

Purchases per day (avg): 0.03 products/day

Purchasing Months (avg): 4 months  

-------------------------------------------------

<b>CLUSTER 3: </b>

Total: 1034 customers (23.99% of total customers)

Spent (avg): $ 720.87

Recency (avg): 94 days

Purchase (avg): 2 purchases

Purchases per day (avg): 0.07 products/day

Purchasing Months (avg): 2 months  

-------------------------------------------------

<b>CLUSTER 0: </b>

Total: 1398 customers (32.43% of total customers)

Spent (avg): $ 371.67

Recency (avg): 153 days

Purchase (avg): 1 purchases

Purchases per day (avg): 0.1 products/day

Purchasing Months (avg): 1 months  

# 10.0 Model Deploy

In [53]:
df94.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4310 entries, 0 to 4365
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   CustomerID        4310 non-null   int64  
 1   GrossRevenue      4310 non-null   float64
 2   RecencyDays       4310 non-null   int64  
 3   Frequency         4310 non-null   float64
 4   ReturnsNo         4310 non-null   float64
 5   QtItems           4310 non-null   int64  
 6   ProductReturns    4310 non-null   float64
 7   PurchasesNo       4310 non-null   float64
 8   ProductsNo        4310 non-null   float64
 9   PurchasingMonths  4310 non-null   int64  
 10  Clusters          4310 non-null   int32  
dtypes: float64(6), int32(1), int64(4)
memory usage: 387.2 KB


In [54]:
# changing data types
df94['ReturnsNo'] = df94['ReturnsNo'].astype(int)
df94['ProductReturns'] = df94['ProductReturns'].astype(int)
df94['PurchasesNo'] = df94['PurchasesNo'].astype(int)
df94['ReturnsNo'] = df94['ReturnsNo'].astype(int)

# 11.0 Database inserting

In [55]:
# saving csv to insert
df94.to_csv('C:/Users/PICHAU/repos/cluster_project/data/veighers_db2.csv')

In [58]:
# saving in json to insert into mongodb
dt_o = json.dumps(df94.to_dict(orient='records'))

In [59]:
# connecting with database
db_url = 'mongodb+srv://felipejaguiar:cluster0@cluster0.rbbckio.mongodb.net/?retryWrites=true&w=majority'
client = MongoClient(db_url)
db = client['fidelity_program_veigh']
coll = db['clusters']

# excluding 
# coll.delete_many({})

# Carregamento dos novos documentos na coleção
novos_dados = json.loads(dt_o)
coll.insert_many(novos_dados)

# Fechamento da conexão com o banco de dados
client.close()